intake/source/base.py

```diff
@@ -53,7 +52,133 @@ class NoEntry(AttributeError):
     pass
 
 
-class DataSource(DictSerialiseMixin):
+class CacheMixin:
+    _cache = None
+
+    @property
+    def cache_dirs(self):
+        return [c._cache_dir for c in self.cache]
+
+    def set_cache_dir(self, cache_dir):
+        for c in self.cache:
+            c._cache_dir = make_path_posix(cache_dir)
+
+    @property
+    def cache(self):
+        if self._cache is None:
+            self._cache = make_caches(self.name, self.metadata.get('cache'),
+                                      catdir=self.metadata.get('catalog_dir',
+                                                             None),
+                                      storage_options=self.storage_options)
+        return self._cache
+
+    @cache.setter
+    def cache(self, csh):
+        self._cache = csh
+
+    def _get_cache(self, urlpath):
+        if len(self.cache) == 0:
+            return [urlpath]
+        return [c.load(urlpath) for c in self.cache]
+
+
+class GraphicsMixin:
+    @property
+    def plots(self):
+        """List custom associated quick-plots """
+        return list(self.metadata.get('plots', {}))
+
+    @property
+    def gui(self):
+        """Source GUI, with parameter selection and plotting"""
+        return self.entry.gui
+
+    @property
+    def plot(self):
+        """
+        Returns a hvPlot object to provide a high-level plotting API.
+
+        To display in a notebook, be sure to run ``intake.output_notebook()``
+        first.
+        """
+        try:
+            from hvplot import hvPlot
+        except ImportError:
+            raise ImportError("The intake plotting API requires hvplot."
+                              "hvplot may be installed with:\n\n"
+                              "`conda install -c pyviz hvplot` or "
+                              "`pip install hvplot`.")
+        metadata = self.metadata.get('plot', {})
+        fields = self.metadata.get('fields', {})
+        for attrs in fields.values():
+            if 'range' in attrs:
+                attrs['range'] = tuple(attrs['range'])
+        metadata['fields'] = fields
+        plots = self.metadata.get('plots', {})
+        return hvPlot(self, custom_plots=plots, **metadata)
+
+    @property
+    def hvplot(self):
+        """
+        Returns a hvPlot object to provide a high-level plotting API.
+        """
+        return self.plot
+
+
+class PersistMixin:
```

Comment:
As part of this I think the `s.get_persisted()` logic in `CatalogEntry.__call__` should be delegated to the `get` methods of the concrete implementations.  In addition data sources being able to opt in/out of implementing the persist logic, I think it is critical that the catalogs also be able to opt out of persist being used.

I guess another option is to do this as a mix-in on the `CatalogEntry` as well?

Reply:
Do you think it's reasonable for the catalog to use persisiting, if the mixin is there? It could even check by isinstance specifically.

Reply:
I added it...